pkg/resourcevalidator/webhook.go

```diff
@@ -0,0 +1,241 @@
+	}
+
+	if !spv.enableResourceValidation {
+		return admission.Allowed("")
+	}
+
+	switch req.Operation {
+	case admissionv1.Create:
+		return spv.HandleCreate(ctx, req, shadowpod, nsClusterID, pointer.BoolDeref(req.DryRun, false))
+	case admissionv1.Delete:
+		return spv.HandleDelete(ctx, req, shadowpod, nsClusterID, pointer.BoolDeref(req.DryRun, false))
+	default:
+		return admission.Denied("Unsupported operation")
+	}
+}
+
+// HandleCreate is the function in charge of handling Creation requests.
+//nolint:gocritic // the signature of this method is imposed by controller runtime.
+func (spv *ShadowPodValidator) HandleCreate(ctx context.Context, req admission.Request, shadowpod *vkv1alpha1.ShadowPod,
+	clusterID string, dryRun bool) admission.Response {
+	spQuota := getQuotaFromShadowPod(shadowpod)
+
+	shadowpodlog.Info(fmt.Sprintf("\tShadowPod %s decoded: UID: %s - clusterID %s", shadowpod.Name, shadowpod.GetUID(), clusterID))
+
+	// Check existence and get resource offer by Cluster ID label
+
+	resourceoffer, err := liqogetters.GetResourceOfferByLabel(ctx, spv.Client, corev1.NamespaceAll,
+		liqolabels.LocalLabelSelector(clusterID))
+	//resourceoffer, err := spv.getResourceOfferByLabel(ctx, clusterID)
+	if err != nil {
+		error := fmt.Errorf("error getting resource offer by label: %w", err)
+		resourceofferlog.Info(error.Error())
+		return admission.Errored(http.StatusInternalServerError, error)
+	}
+
+	// Get ResourceOffer Quota
+	roQuota := getQuotaFromResourceOffer(resourceoffer)
+
+	resourceofferlog.Info(fmt.Sprintf("ResourceOffer found for clusterID %s with %s ", clusterID, quotaFormatter(roQuota, "Quota")))
+
+	peeringInfo := getOrCreatePeeringInfo(spv.PeeringCache, clusterID, resourceoffer.OwnerReferences[0].Name, roQuota)
+
+	spd, found := peeringInfo.getShadowPodDescription(shadowpod.GetName())
+	if !found {
+		// Create a new ShadowPodDescription for caching purposes
+		spd = createShadowPodDescription(shadowpod.GetName(), string(shadowpod.GetUID()), spQuota)
+	} else if spd.running {
+		return admission.Denied("ShadowPod already exists")
+	}
+
+	shadowpodlog.Info(quotaFormatter(spd.getQuota(), "\tShadowPod resource limits"))
+
+	err = peeringInfo.testAndUpdatePeeringInfo(spd, admissionv1.Create, dryRun)
+	if err != nil {
+		return admission.Denied(err.Error())
+	}
+
+	return admission.Allowed("")
+}
+
+// HandleUpdate is the function in charge of handling Update requests.
+//nolint:gocritic // the signature of this method is imposed by controller runtime.
+func (spv *ShadowPodValidator) HandleUpdate(req admission.Request) admission.Response {
+	obj := req.Object
+	oldObj := req.OldObject
+	shadowpod, err := spv.DecodeShadowPod(obj)
+	oldShadowpod, oldErr := spv.DecodeShadowPod(oldObj)
+	if err != nil || oldErr != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	if pod.CheckShadowPodUpdate(&shadowpod.Spec.Pod, &oldShadowpod.Spec.Pod) {
+		return admission.Allowed("")
+	}
```

Comment:
You should ensure that the cluster ID label does not get changed (you should be able to reuse the common function extracted from the Handle logic above).